src/dataflow/src/source/loki.rs

```diff
@@ -41,128 +38,142 @@ impl LokiSourceReader {
-        })
+            .await
+            .context("Connection error")?
+            .error_for_status()
+            .context("HTTP error")?
+            .bytes()
+            .await
+            .context("Download error")?;
+        let result: QueryResult = serde_json::from_slice(&response).context("Deserialize error")?;
+        let Data::Streams(streams) = result.data;
+
+        #[derive(Debug, Serialize)]
+        struct LokiRow<'a> {
+            timestamp: &'a str,
+            line: &'a str,
+            labels: &'a HashMap<&'a str, &'a str>,
+        }
+
+        // TODO(bsull): we could get rid of this intermediate Vec if we handled the timestamp sending
+        // in this function instead, but for now it's quite nice to be able to see the resulting JSON
+        // in a test.
+        let lines: Vec<String> = streams
+            .iter()
+            .flat_map(|s| {
+                s.values.iter().map(|v| {
+                    serde_json::to_string(&LokiRow {
+                        timestamp: v.ts,
+                        line: v.line,
+                        labels: &s.labels,
+                    })
+                    .expect("Loki data should be valid JSON")
+                })
+            })
+            .collect();
+        Ok(lines)
     }
 }
 
 #[async_trait]
 impl SimpleSource for LokiSourceReader {
     async fn start(mut self, timestamper: &Timestamper) -> Result<(), SourceError> {
-        let stream = self.new_stream();
-        tokio::pin!(stream);
-
-        while let Some(entry) = stream.next().await {
-            match entry {
-                Ok(result) => {
-                    let Data::Streams(streams) = result.data;
-
-                    #[derive(Debug, Serialize)]
-                    struct LokiRow<'a> {
-                        line: &'a str,
-                        labels: &'a HashMap<String, String>,
-                    }
-
-                    let lines: Vec<String> = streams
-                        .iter()
-                        .flat_map(|s| {
-                            s.values.iter().map(|v| {
-                                serde_json::to_string(&LokiRow {
-                                    line: &v.line,
-                                    labels: &s.labels,
-                                })
-                                .unwrap()
-                            })
-                        })
-                        .collect();
-
-                    for line in lines {
-                        let row = Row::pack_slice(&[Datum::String(&line)]);
-
-                        timestamper.insert(row).await.map_err(|e| SourceError {
-                            source_id: self.source_id.clone(),
-                            error: SourceErrorDetails::FileIO(e.to_string()),
-                        })?;
+        let mut interval = tokio::time::interval(self.batch_window);
+        loop {
+            interval.tick().await;
+            match self.tick().await {
```

Comment:
It's more or less equivalent to doing:

```
        let mut intervals = IntervalStream::new(tokio::time::interval(self.batch_window));
        while intervals.next().await.is_some() {
            match self.tick().await {
                ...
```

I think. The main thing here is that we've got rid of the stream and handle everything related to the response in `LokiSourceReader::tick`, so that we don't need to return any owned data (which would require allocating) - we can just deserialize using the borrowed bytes of the HTTP response.